### ISSUE TITLE
Softmax

### ISSUE THREAD

**Issue by tcpipchip** [ultralytics/ultralytics#23823]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Yolo works perfectly to create .pt and i convert to onnx to cvimodel (tpu mlir) (sg2002) to use as OBJECT DETECTOR that works perfectly the derection.

To CLASSIFICATION , yolo converts to pt perfectly  and to onnx. But when i convert to cvimodel, fails the sg2002 cvimodel execution of classification. Always my inference returns 0 class.

Is there some way to remove the LAYER SOFTMAX using YOLO ?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @tcpipchip, thank you for your interest in Ultralytics 🚀! This is an automated response to help move things forward quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 conversion/runtime issue, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so the team can debug effectively. Including your exact export/convert commands, model variant, environment details, and full logs will help a lot 🙏

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You can patch the head to remove softmax:


```python

import torch
import types

def raw_forward(self, x):
    """Modified forward pass returning raw logits."""
    if isinstance(x, list):
        x = torch.cat(x, 1)
    x = self.linear(self.drop(self.pool(self.conv(x)).flatten(1)))
    # Return raw logits x instead of softmax(x)
    return x if self.export else (x, x)

# Apply the patch to the classification head
model.model.model[-1].forward = types.MethodType(raw_forward, model.model.model[-1])

model.export(format="onnx")
```

**Comment by tcpipchip** [comment]
I will try on tdl sdk sophgo sg2002 on monday. Thanks

**Comment by glenn-jocher** [comment]
Sounds good — that patch targets the right spot, since the classification head normally applies `softmax()` at inference/export in [`Classify.forward()`](https://docs.ultralytics.com/reference/nn/modules/head/); if the SG2002 still predicts only class `0`, please share your exact `model.export(...)` args, TPU-MLIR conversion command, `yolo checks`, and the full text runtime logs so we can compare the ONNX logits against the `cvimodel` output.

**Comment by tcpipchip** [comment]
Hi, to remove the SOFTMAX, i executed

import torch
import types
from ultralytics import YOLO

model = YOLO("best.pt")

def raw_forward(self, x):
    if isinstance(x, list):
        x = torch.cat(x, 1)
    x = self.linear(self.drop(self.pool(self.conv(x)).flatten(1)))
    return x if self.export else (x, x)

model.model.model[-1].forward = types.MethodType(raw_forward, model.model.model[-1])

model.export(format="onnx")

and then

model_transform.py \
    --model_name yolov8n_cls \
    --model_def best.onnx \
    --input_shapes [[1,3,640,640]] \
	--mean "0,0,0" \
	--scale "0.00392156862745098,0.00392156862745098,0.00392156862745098" \
	--keep_aspect_ratio \
	--pixel_format rgb \
	--channel_format nchw \
	--test_input test.jpg \
	--tolerance 0.99,0.99 \
	--test_result yolov8n_cls_top_outputs.npz \
    --mlir yolov8n_cls.mlir

run_calibration.py yolov8n_cls.mlir \
  --dataset cores \
  --input_num 100 \
  -o yolo_cls_cali_table	


model_deploy.py \
    --mlir yolov8n_cls.mlir \
    --quantize INT8 \
	--quant_input \
    --calibration_table yolo_cls_cali_table \
    --processor cv181x \
	--test_input yolov8n_cls_in_f32.npz \
    --test_reference yolov8n_cls_top_outputs.npz \
	--tolerance 0.9,0.6 \
    --model yolo_cls_cv181x_int8_sym.cvimodel  

[_data_.zip](https://github.com/user-attachments/files/25829087/_data_.zip)

[yolo_cls_cv181x_int8_sym.zip](https://github.com/user-attachments/files/25829047/yolo_cls_cv181x_int8_sym.zip)

With officials classify cvimodel works...

**Comment by tcpipchip** [comment]
Please...wait for news! Dont open those files.

**Comment by tcpipchip** [comment]
go on...

**Comment by tcpipchip** [comment]
I followed the instructions o MAIXCAM, that is the same processor of mine.

They said that for classification, remove the SOFTMAX...

This morning i trainned with their maixhub and got this cvimodel that works nice to me...

https://wiki.sipeed.com/maixpy/doc/en/ai_model_converter/maixcam.html

[model-254417.maixcam.zip](https://github.com/user-attachments/files/25839897/model-254417.maixcam.zip)

The problem to investigate is that they generate directly the cvimodel

here the car colors class trainned 

https://1drv.ms/u/c/36e983d4fed6fe4f/IQAFnM44hwdoRYzBsBKDIdb7AXGxIBwZRv_dRQa3tyrTAKw?e=kj6sny

they use backbone mobilenet_v2

<img width="1600" height="860" alt="Image" src="https://github.com/user-attachments/assets/fa423dca-ae60-4e79-afcc-43b41c32ef17" />

**Comment by tcpipchip** [comment]
Finally works CLASSIFY on SG2002!

**Comment by glenn-jocher** [comment]
Good to hear it’s working on SG2002 now; if you can share the exact change that fixed it, that would help other users with the same classification export pipeline.

**Comment by tcpipchip** [comment]
Finally working nice!

```
[root@milkv-duo]~# ./sample_img_classification CLS_YOLOV8 /root /root/amarelo.jp
g
[tdl_model_factory.cpp:80] [I] input model config file is empty, load model config from /mnt/system/configs/model/model_factory.json
[tdl_model_factory.cpp:104] [I] load model config from /mnt/system/configs/model/model_factory.json done,model size:66
[tdl_model_factory.cpp:226] [I] setModelDir success,model_dir:/root
[tdl_model_factory_internal.cpp:40] [I] getModelInstance model_type:CLS_YOLOV8
[tdl_model_factory.cpp:412] [I] model_path: /root/cv181x/yolo_cls_cv181x_int8_sym_cv181x.cvimodel
pred_label: 8
[root@milkv-duo]~#
```

```
import torch
import types
from ultralytics import YOLO

model = YOLO("best.pt")

def raw_forward(self, x):
    """Modified forward pass returning raw logits."""
    if isinstance(x, list):
        x = torch.cat(x, 1)
    x = self.linear(self.drop(self.pool(self.conv(x)).flatten(1)))
    # Return raw logits x instead of softmax(x)
    return x if self.export else (x, x)

model.model.model[-1].forward = types.MethodType(raw_forward, model.model.model[-1])
```
```
import onnx

onnx_model = onnx.load("best.onnx")

for node in onnx_model.graph.node:
    for i, inp in enumerate(node.input):
        if inp == "images":
            node.input[i] = "input0"

# Renomear no graph input
for inp in onnx_model.graph.input:
    if inp.name == "images":
        inp.name = "input0"

for out in onnx_model.graph.output:
    if out.name != "output0":
        out.name = "output0"

onnx.save(onnx_model, "best_renamed.onnx")
print("Salvo como best_renamed.onnx")
```

`model_transform.py \
    --model_name yolov8n_cls \
    --model_def best.onnx \
    --input_shapes [[1,3,640,640]] \
	--mean "0,0,0" \
	--scale "0.00392156862745098, 0.00392156862745098, 0.00392156862745098" \
	--keep_aspect_ratio \
	--output_names "output0" \
	--pixel_format rgb \
	--channel_format nchw \
	--test_input green.jpg \
	--test_result yolov8n_cls_top_outputs.npz \
	--tolerance 0.95,0.60 \
    --mlir yolov8n_cls.mlir
		
run_calibration.py yolov8n_cls.mlir \
  --dataset testes \
  --input_num 1600 \
  -o yolo_cls_cali_table	
  
model_deploy.py \
    --mlir yolov8n_cls.mlir \
    --quantize INT8 \
	--quant_input \
    --calibration_table yolo_cls_cali_table \
    --processor cv181x \
	--test_input yolov8n_cls_in_f32.npz \
    --test_reference yolov8n_cls_top_outputs.npz \
	--tolerance 0.95,0.60 \
    --model yolo_cls_cv181x_int8_sym.cvimodel
`

**Comment by glenn-jocher** [comment]
Thanks for posting the final working setup — this matches the default [`Classify.forward()` behavior](https://docs.ultralytics.com/reference/nn/modules/head/), which applies `softmax()` at inference/export, so in your SG2002 pipeline the key was exporting raw logits and using explicit ONNX I/O names like `input0` and `output0`; this should help others seeing all-class-`0` outputs with `cvimodel` classification.

**Comment by tcpipchip** [comment]
Wellcome. Have you plans to add on ultralytics platform the option to convert direclty to cvimodel...for example a plug-in?

**Comment by glenn-jocher** [comment]
We do not have direct `cvimodel` export planned right now, but it’s a reasonable feature request; please open a separate request in [Ultralytics Discussions](https://github.com/orgs/ultralytics/discussions) for visibility, and for now the supported path remains the standard [export workflow](https://docs.ultralytics.com/modes/export/) followed by your TPU-MLIR conversion.

**Comment by tcpipchip** [comment]
Thank you

**Comment by glenn-jocher** [comment]
You’re welcome — glad it’s working now, and thanks for sharing the SG2002 workaround here for others.

**Comment by tcpipchip** [comment]
Hi, is there some tip about to use to POSE on recamera ?

**Comment by glenn-jocher** [comment]
That’s a separate topic from this classification thread, so please open a new issue with your exact reCamera model, whether you’re exporting `yolo26n-pose.pt` to ONNX first, `yolo checks`, and the full text logs; for pose the right starting point is the [Ultralytics YOLO pose docs](https://docs.ultralytics.com/tasks/pose/), and reCamera’s documented deployment path goes through ONNX → MLIR → `cvimodel`. ([docs.ultralytics.com](https://docs.ultralytics.com/tasks/pose/))

**Comment by tcpipchip** [comment]
Wow, sorry @glenn-jocher 
Btw, can you add the POSE option on yolov8 platform Ultralytics :)

**Comment by Y-T-G** [comment]
@tcpipchip Thanks for the request. We will add them. You can also request feature or report bugs using the "Feedback" button on the Platform website

**Comment by Y-T-G** [comment]
@tcpipchip YOLOv8 Pose models have been added to Platform

**Comment by tcpipchip** [comment]
@Y-T-G thank you so much!
